src/main/java/software/amazon/smithy/lsp/ServerState.java

```diff
@@ -0,0 +1,236 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+import org.eclipse.lsp4j.WorkspaceFolder;
+import software.amazon.smithy.lsp.document.Document;
+import software.amazon.smithy.lsp.project.Project;
+import software.amazon.smithy.lsp.project.ProjectAndFile;
+import software.amazon.smithy.lsp.project.ProjectFile;
+import software.amazon.smithy.lsp.project.ProjectLoader;
+import software.amazon.smithy.lsp.protocol.LspAdapter;
+import software.amazon.smithy.lsp.util.Result;
+
+/**
+ * Keeps track of the state of the server.
+ *
+ * @param detachedProjects Map of smithy file **uri** to detached project
+ *                         for that file
+ * @param attachedProjects Map of directory **path** to attached project roots
+ * @param workspacePaths Paths to roots of each workspace open in the client
+ * @param managedUris Uris of each file managed by the server/client, i.e.
+ *                    files which may be updated by didChange
+ * @param lifecycleManager Container for ongoing tasks
+ */
+public record ServerState(
+        Map<String, Project> detachedProjects,
+        Map<String, Project> attachedProjects,
+        Set<Path> workspacePaths,
+        Set<String> managedUris,
+        DocumentLifecycleManager lifecycleManager
+) {
+    private static final Logger LOGGER = Logger.getLogger(ServerState.class.getName());
+
+    /**
+     * Create a new, empty server state.
+     */
+    public ServerState() {
+        this(
+                new HashMap<>(),
+                new HashMap<>(),
+                new HashSet<>(),
+                new HashSet<>(),
+                new DocumentLifecycleManager());
+    }
+
+    /**
+     * @param uri Uri of the document to get
+     * @return The document if found and it is managed, otherwise {@code null}
+     */
+    public Document getManagedDocument(String uri) {
+        if (managedUris.contains(uri)) {
+            ProjectAndFile projectAndFile = findProjectAndFile(uri);
+            if (projectAndFile != null) {
+                return projectAndFile.file().document();
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param path The path of the document to get
+     * @return The document if found and it is managed, otherwise {@code null}
+     */
+    public Document getManagedDocument(Path path) {
+        if (managedUris.isEmpty()) {
+            return null;
+        }
+
+        String uri = LspAdapter.toUri(path.toString());
+        return getManagedDocument(uri);
+    }
+
+    ProjectAndFile findProjectAndFile(String uri) {
+        String path = LspAdapter.toPath(uri);
+        // TODO: From isDetached
+        for (Project project : attachedProjects.values()) {
+            ProjectFile projectFile = project.getProjectFile(path);
+            if (projectFile != null) {
+                detachedProjects.remove(uri);
```

Comment:
Im a little confused why we remove it from detached projects here? I would have thought that would get checked correctly in the isDetached below?

Reply:
Yea it does, but I'm not calling that method here. I think I meant to end up removing `isDetached`, but it ended up being too useful. I did some refactoring and pulled the removal logic into the new `findAttachedAndRemoveDetached`